data_pipeline_api/java_api/src/main/java/uk/ramp/api/FileApi.java

```diff
@@ -0,0 +1,98 @@
+import uk.ramp.yaml.YamlFactory;
+
+/** Java implementation of Data Pipeline File API. */
+public class FileApi implements AutoCloseable {
+  private final MetadataSelector metadataSelector;
+  private final AccessLogger accessLogger;
+  private final OverridesApplier overridesApplier;
+  private final FileDirectoryNormaliser fileDirectoryNormaliser;
+
+  public FileApi(Path configFolderPath) {
+    this(Clock.systemUTC(), configFolderPath);
+  }
+
+  FileApi(Clock clock, Path parentPath) {
+    var hasherFactory = new HasherFactory();
+    var metadataSelectorFactory = new MetadataSelectorFactory();
+    var configFactory = new ConfigFactory();
+    var accessLoggerFactory = new AccessLoggerFactory();
+    var yamlFactory = new YamlFactory();
+    var openTimestamp = clock.instant();
+    this.fileDirectoryNormaliser = new FileDirectoryNormaliser(parentPath.toString());
+    var config =
+        configFactory.config(
+            yamlFactory.yamlReader(),
+            hasherFactory,
+            new FileReader(),
+            openTimestamp,
+            fileDirectoryNormaliser);
+    this.overridesApplier = new OverridesApplier(config);
+    this.accessLogger =
+        accessLoggerFactory.accessLogger(config, yamlFactory.yamlWriter(), clock, openTimestamp);
+    this.metadataSelector =
+        metadataSelectorFactory.metadataSelector(
+            hasherFactory.contentsHasher(),
+            yamlFactory.yamlReader(),
+            config.failOnHashMisMatch().orElse(true),
+            fileDirectoryNormaliser);
+  }
+
+  /**
+   * Return file for reading corresponding to the given metadata. The file contents are hashed, and
+   * a record is made of the read.
+   *
+   * @return FileChannel for input
+   * @param query input query
+   */
+  public FileChannel openForRead(MetadataItem query) throws IOException {
+    var overridenQuery = overridesApplier.applyReadOverrides(query);
+    var metaDataItem = metadataSelector.find(overridenQuery);
+    var filename = metaDataItem.filename().orElseThrow();
+    var normalisedFilename = fileDirectoryNormaliser.normalisePath(filename);
+    accessLogger.logRead(query, metaDataItem);
+    return FileChannel.open(Path.of(normalisedFilename), READ);
+  }
+
+  /**
+   * Return matching file for update corresponding to the given metadata. When the file is closed
+   * the file contents are hashed, and a record is made of the write.
+   *
+   * @return FileChannel for output
+   * @param query input query
+   */
+  public FileChannel openForWrite(MetadataItem query) throws IOException {
+    var overridenQuery = overridesApplier.applyWriteOverrides(query);
+    var filename = overridenQuery.filename().orElseThrow();
+    var normalisedFilename = fileDirectoryNormaliser.normalisePath(filename);
+    Files.createDirectories(Path.of(fileDirectoryNormaliser.parentPath()));
+    Files.createFile(Path.of(normalisedFilename));
+    accessLogger.logWrite(query, overridenQuery);
```

Comment:
I've only just scanned this, and I'm not very familiar with modern Java so maybe I've missed something, but does the write logging happen _before_ the write, or once the file has been closed? It should be the latter, and the hash of the resulting file should be computed. Again, apologies if I've just missed something.

Reply:
A very good question and well spotted. You're correct. The 'write logging' happens before the write to file actually happens. We log that there has been a write access just before returning the file handle. 

I didn't realise that this was a strict requirement. This will require some fiddling around as Java doesn't let you directly override the close() method on the file handle (FileChannel objects). 

The point you mentioned about the hash of the resulting file needing to be computed again is important. 

I'll work on getting this functionality implemented. 

Thanks

Reply:
Yeah nw, it's probably the most subtle/awkward part of how this works, which is why I went looking for it. I still have to look through the rest, but I imagine it's basically fine.